core/src/test/java/com/airbnb/aerosolve/core/util/ReinforcementLearningTest.java

```diff
@@ -0,0 +1,181 @@
+    return dict;
+  }
+  
+  public HashMap makeState(double x, double y) {
+    HashMap stateFeatures = new HashMap<Double, String>();
+    stateFeatures.put("x", x);
+    stateFeatures.put("y", y);
+    return stateFeatures;
+  }
+  
+  public HashMap makeAction(double dx, double dy) {
+    HashMap actionFeatures = new HashMap<Double, String>();
+    actionFeatures.put("dx", dx);
+    actionFeatures.put("dy", dy);
+    return actionFeatures;
+  }
+
+  public FeatureVector makeFeatureVector(HashMap state, HashMap action) {
+    HashMap floatFeatures = new HashMap<Double, HashMap<Double, String>>();
+    floatFeatures.put("S", state);
+    floatFeatures.put("A", action);
+    FeatureVector featureVector = new FeatureVector();
+    featureVector.setFloatFeatures(floatFeatures);
+    return featureVector;
+  }
+  
+  public ArrayList<FeatureVector> stateActions(double x, double y) {
+    HashMap currState = makeState(x, y);
+    ArrayList<FeatureVector> potential = new ArrayList<>();
+    HashMap up = makeAction(0.0f, 1.0f);
+    HashMap down = makeAction(0.0f, -1.0f);
+    HashMap left = makeAction(-1.0f, 0.0f);
+    HashMap right = makeAction(1.0f, 0.0f);
+
+    potential.add(makeFeatureVector(currState, up));
+    potential.add(makeFeatureVector(currState, down));
+    potential.add(makeFeatureVector(currState, left));
+    potential.add(makeFeatureVector(currState, right));
+    return potential;
+  }
+
+  AbstractModel makeModel() {
+    KernelModel model = new KernelModel();
+    StringDictionary dict = makeDictionary();
+    model.setDictionary(dict);
+    List<SupportVector> supportVectors = model.getSupportVectors();
+    Random rnd = new Random(12345);
+    for (double x = 0.0; x <= 10.0; x += 1.0) {
+      for (double y = 0.0; y <= 10.0; y += 1.0) {
+        ArrayList<FeatureVector> potential = stateActions(x, y);
+        for (FeatureVector sa : potential) {
+          FloatVector vec = dict.makeVectorFromSparseFloats(Util.flattenFeature(sa));
+          SupportVector sv = new SupportVector(vec, FunctionForm.RADIAL_BASIS_FUNCTION, 1.0f, 0.0f);
+          supportVectors.add(sv);
+        }
+      }
+    }
+    return model;
+  }
+
+  // In the cliffworld scenario, the agent starts at (0,0)
+  // There is a cliff in the region (2.0, 0.0) to (8.0, 2.0) if the agent steps in the cliff it gets a reward of -100 and teleports back to the start.
+  // Otherwise the reward is -1. If it hits the goal the reward is 10.0
+  //
+  //
+  //
+  //
+   //  Start    CLIFF    Goal
+  //   (0,0)    CLIFF    (10, 0)
+  public float runEpisode(AbstractModel model, int epoch, Random rnd) {
+    float sum = 0.0f;
+    double x = 0.0;
+    double y = 0.0;
+    FeatureVector prevStateAction = null;
+    float prevReward = 0.0f;
+    FeatureVector currStateAction = null;
+    float epsilon = 0.1f;
+    float learningRate = 0.1f;
+    if (epoch > 90) {
+      epsilon = 0.0f;
+      learningRate = 0.001f;
+    }
+    boolean done = false;
+    while (!done) {
```

Comment:
I think this works, but it might be a little easier to follow if a full SARSA update is done at every iteration (as opposed to deferring it one period).

The revised approach would be:
1. Initialize currStateAction with x=0, y=0, pick initial action
2. For each iteration:
  1. Evaluate reward, next (x, y) based on currStateAction
  2. If (x, y) is not terminal, pick greedy action for next x, next y and set in nextStateAction; otherwise, nextStateAction is null
  3. Do full SARSA update using (currStateAction, reward, nextStateAction)
  4. Set currStateAction to nextStateAction